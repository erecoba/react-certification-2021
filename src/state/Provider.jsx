import React, { createContext, useContext, useReducer } from 'react';

import youtubeReducer, {
  initialState as youtubeInitialState,
} from './reducers/youtube/youtube';
import generalReducer, { initialState as generalInitialState } from './reducers/general';
import sessionReducer, { initialState as sessionInitialState } from './reducers/session';

const YoutubeContext = createContext({ ...youtubeInitialState });
const GeneralContext = createContext({ ...generalInitialState });
const SessionContext = createContext({ ...sessionInitialState });

const useYoutubeVideo = () => {
  const context = useContext(YoutubeContext);
  if (!context) {
    throw new Error("Can't use useYoutubeVideo with out a Provider");
  }
  return context;
};

const useGeneral = () => {
  const context = useContext(GeneralContext);
  if (!context) {
    throw new Error("Can't use useGeneral with out a Provider");
  }
  return context;
};

const useSession = () => {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error("Can't use useSession with out a Provider");
  }
  return context;
};

const Provider = ({ children }) => {
  const [youtubeState, youtubeDispatch] = useReducer(youtubeReducer, youtubeInitialState);
  const [generalState, generalDispatch] = useReducer(generalReducer, generalInitialState);
  const [sessionState, sessionDispatch] = useReducer(sessionReducer, sessionInitialState);

  return (
    <GeneralContext.Provider value={{ generalState, generalDispatch }}>
      <YoutubeContext.Provider value={{ youtubeState, youtubeDispatch }}>
        <SessionContext.Provider value={{ sessionState, sessionDispatch }}>
          {children}
        </SessionContext.Provider>
      </YoutubeContext.Provider>
    </GeneralContext.Provider>
  );
};

export { useYoutubeVideo, useGeneral, useSession };

export default Provider;
