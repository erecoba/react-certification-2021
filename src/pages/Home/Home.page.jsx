import React from 'react';
import Header from '../../components/Header';
import Card from '../../components/Card';
import { Container } from './Home.styled';

import { useYoutubeVideo } from '../../state/Provider';

function HomePage() {
  const { youtubeState } = useYoutubeVideo();
  const { videos } = youtubeState;

  return (
    <>
      <Header />
      <Container>
        {videos.map((video) => (
          <Card
            key={video.id}
            id={typeof video.id === 'object' ? video.id.videoId : video.id}
            title={video.snippet.title}
            description={video.snippet.description}
            imgUrl={video.snippet.thumbnails.medium.url}
            channelTitle={video.snippet.channelTitle}
          />
        ))}
      </Container>
    </>
  );
}

export default HomePage;
