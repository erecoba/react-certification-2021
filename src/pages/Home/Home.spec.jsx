import React from 'react';
import { act, render } from '@testing-library/react';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import Provider from '../../state/Provider';
import Home from './Home.page';

import MockVideosPopular from '../../mocks/videos-popular.json';
import MockVideosSearch from '../../mocks/videos-searching.json';
import MockVideosRelated from '../../mocks/videos-related.json';
import MockVideosDetail from '../../mocks/video-detail.json';

jest.mock('../../services', () => {
  return jest.fn(() => ({
    youtube: {
      getVideos: jest.fn(() => MockVideosPopular),
      searchVideos: jest.fn(() => MockVideosSearch),
      getRelatedVideos: jest.fn(() => MockVideosRelated),
      getStatistics: jest.fn(() => MockVideosDetail),
    },
  }));
});

describe('Home Page', () => {
  let theme;
  let history;

  beforeEach(() => {
    history = createMemoryHistory();
    theme = {
      base: '#f0f',
      background: '#fff',
      label: '#000',

      inputHoverBorder: '#f0f',
      inputHoverBg: '#ff0',

      colors: {
        white: '#fff',
      },
    };
  });

  it('should render correctly Home', async () => {
    let wrapper;

    await act(async () => {
      wrapper = render(
        <ThemeProvider theme={theme}>
          <Provider>
            <Router history={history}>
              <Home />
            </Router>
          </Provider>
        </ThemeProvider>
      );
    });

    let fragment;
    await act(async () => {
      fragment = wrapper.asFragment(
        <ThemeProvider>
          <Provider>
            <Router history={history}>
              <Home />
            </Router>
          </Provider>
        </ThemeProvider>
      );
    });

    expect(fragment).toMatchSnapshot();
  });
});
