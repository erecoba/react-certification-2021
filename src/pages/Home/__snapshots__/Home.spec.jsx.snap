// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render correctly inside button: main 1`] = `"DIV"`;

exports[`Avatar should render correctly inside button: nav 1`] = `"DIV"`;
